File: sorting/insertion-sort/InsertionSort.py
from sys import argv

def insertionSort(array, N):
    for i in range(1, N):
        for j in range(i, 0, -1):
            if array[j] < array[j-1]:
                swap(array, j, j-1)

def insertionSort2(array, N):
    for i in range(1, N):
        x = array[i]
        j = i
        while j > 0 and x < array[j-1]:
            array[j] = array[j-1]
            j -= 1
        array[j] = x

def swap(A, i, j):
  temp = A[i]
  A[i] = A[j]
  A[j] = temp

# Prepare memory
N = len(argv) - 1
array = []
array2 = []

# Read integers
for x in argv[1:]:
    I = int(x)
    array.append(I)
    array2.append(I)

# Demonstrate
print("insertionSort")
insertionSort(array, N)
print(array)

print("insertionSort2")
insertionSort2(array2, N)
print(array2)

File: sorting/insertion-sort/InsertionSort.java
public class InsertionSort {

    public static void insertionSort(int[] array, final int N) {
        for (int i=1; i < N; i++)
            for (int j=i; j > 0; j--)
                if (array[j] < array[j-1])
                    swap(array, j, j-1);
    }

    public static void insertionSort2(int array[], final int N) {
        for (int i=1; i < N; i++) {
            int x = array[i];
            int j = i;
            while ( (j > 0) && (x < array[j-1]) ) {
                array[j] = array[j-1];
                j--;
            }
            array[j] = x;
        }
    }

    public static void main(String[] args) {

        // Prepare memory
        final int N = args.length;
        int[] array = new int[N],
            array2 = new int[N];

        // Read integers
        for (int i=0; i < N; i++) {
            final int I = Integer.parseInt(args[i]);
            array[i] = I;
            array2[i] = I;
        }

        // Demonstrate
        System.out.println("insertionSort");
        insertionSort(array, N);
        print(array, N);

        System.out.println("insertionSort2");
        insertionSort2(array2, N);
        print(array2, N);
    }

    private static void print(int array[], final int N) {
        System.out.print('[');
        for (int i=0; i < N; i++)
            System.out.printf(i < N-1 ? "%d, "
                : "%d]\n", array[i]);
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}